src/core/ddsi/include/dds/ddsi/ddsi_deadline.h

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright(c) 2006 to 2019 ADLINK Technology Limited and others
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0, or the Eclipse Distribution License
+ * v. 1.0 which is available at
+ * http://www.eclipse.org/org/documents/edl-v10.php.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR BSD-3-Clause
+ */
+#ifndef DDSI_DEADLINE_H
+#define DDSI_DEADLINE_H
+
+#include "dds/ddsrt/circlist.h"
+#include "dds/ddsi/q_time.h"
+#include "dds/ddsi/q_globals.h"
+#include "dds/ddsi/q_xevent.h"
+
+#if defined (__cplusplus)
+extern "C" {
+#endif
+
+typedef nn_mtime_t (*deadline_missed_cb_t)(void *hc, nn_mtime_t tnow);
+
+struct deadline_adm {
+  struct ddsrt_circlist list;               /* linked list for deadline missed */
+  struct xevent *evt;                       /* xevent that triggers when deadline expires for an instance */
+  deadline_missed_cb_t deadline_missed_cb;  /* callback for deadline missed; this cb can use deadline_next_missed_locked to get next instance that has a missed deadline */
+  size_t list_offset;                       /* offset of deadline_adm element in whc or rhc */
+  size_t elem_offset;                       /* offset of deadline_elem element in whc or rhc instance */
+  ddsi_duration_t dur;                      /* deadline duration */
```

Comment:
A ``ddsi_duration_t`` matches the wire format (fixed-point with a resolution of 2^-32s) and forces conversions left and right. It'd be better to represent it as nanoseconds throughout.